contracts/BoostedMasterChef.sol

```diff
@@ -381,17 +470,48 @@ contract MasterChefJoeV3 is Ownable, ReentrancyGuard {
         updatePool(pid);
         PoolInfo memory pool = poolInfo[pid];
         UserInfo storage user = userInfo[pid][msg.sender];
+        uint256 userBoost = getUserBoost(pid, msg.sender);
 
         if (user.amount > 0) {
             // Harvest JOE
-            uint256 pending = user.amount.mul(pool.accJoePerShare).div(ACC_TOKEN_PRECISION).sub(user.rewardDebt);
+
+            uint256 pending = user
+                .amount
+                .mul(userBoost)
+                .mul(pool.accJoePerShare)
+                .div(ACC_TOKEN_PRECISION)
+                .sub(user.rewardDebt)
+                .add(claimableJoe[pid][msg.sender]);
+            claimableJoe[pid][msg.sender] = 0;
             JOE.safeTransfer(msg.sender, pending);
             emit Harvest(msg.sender, pid, pending);
         }
 
         // Effects
+
+        // If the user has left the pool altogether we subtract all their
+        // veJoe balance from `pool.totalVeJoe`, and full boosted balance
+        // from `pool.totalBoostedAmount` otherwise we just adjust down the
+        // total boosted amount as total vejoe is unchanged.
+
+        // Update the `pool.totalBoostedAmount`.
+        if (user.amount.sub(amount) == 0) {
+            pool.totalVeJoe = pool.totalVeJoe.sub(VEJOE.balanceOf(msg.sender));
```

Comment:
Why not subtract with `userInfo.veJoeBalance` here? What if user increases his veJOE balance since it was updated in BCMJ?